File: Snowman.m
clear;
%Grid Variables
gridSize = 200;

global frostGrid

frostGrid = 2*ones(gridSize,gridSize);

originX = ceil(gridSize/2);
originY = ceil(gridSize/2);

%Drying Probability Variables
R = 2^(0.5);
C = 1;

%Set seed
frostGrid(originX,originY) = 1;

global walkerRadius
walkerRadius = 30;
radiusDelta = 1;
walkerTheta = 0;
walkerdTheta = 2*pi/360;

global running
running = true;


%Main Loop
while running
    %Get Initial Walker Point
    walkerTheta = rand()*4*pi;
    walkerX = floor(cos(walkerTheta)*walkerRadius) + originX;
    walkerY = floor(sin(walkerTheta)*walkerRadius) + originY;

    %Walk Walker
    walk(walkerX,walkerY,originX,originY,gridSize,radiusDelta);
    
    %img = normalize(frostGrid,'range',[0 1]);
   % imshow(img,'InitialMagnification',800)
    plt = pcolor(frostGrid)
    colormap cool
    shading interp
    pause(1e-16)
    plt.LineWidth = 0.1;

    
    %Advanace Theta
  %  walkerTheta = walkerTheta + walkerdTheta;
end

walkerTheta = 0;
while walkerTheta < 2*pi
    x = floor(cos(walkerTheta)*walkerRadius) + originX;
    y = floor(sin(walkerTheta)*walkerRadius) + originY;
    frostGrid(x,y) = 3;
    walkerTheta = walkerTheta + walkerdTheta;
end



function out = walk(x,y,ox,oy,gridSize,radiusDelta,C)
    global frostGrid
    global walkerRadius
    global running
    
    originalX = x;
    originalY = y;
    
    maxRestart = 20;
    restartCount = 0;
    %While not attached Loop
    while true
        %Move Walker randomly
        direction = rand();
        if direction < 0.25
            x = x + 1;
        elseif direction < 0.5
            y = y + 1;
        elseif direction < 0.75
            x = x - 1;
        elseif direction < 1
            y = y -1;
        end

        %Check if walker is in boundry
        if ((x-ox)^2 + (y-oy)^2) > walkerRadius^2 || frostGrid(x,y) == 0 || frostGrid(x,y) == 1
            %Ignore Walker, Move on to next
            out = -1;
            if restartCount < maxRestart
                x = originalX;
                y = originalY;
            else
                break;
            end
            
            restartCount = restartCount + 1;
            %fprintf("OUT OF BOUND\n");
        end

        %Walker is in boundary, Check for frosting
        didFrost = frostProbability(x,y);
        compareProb = rand();
        if didFrost > compareProb           
            frostGrid(x,y) = 1;
            
            distanceToRadius = pdist([ox,oy;x,y],'euclidean');
            
            if distanceToRadius >= 0.8*walkerRadius
                if walkerRadius<(gridSize/2-2)
                    walkerRadius = walkerRadius + radiusDelta;
                   % fprintf("radius: " + walkerRadius +"\n");
                else
                    running = false
                end
            end
%            if distanceToRadius >= gridSize/2 -2
 %               running = false;
  %          end
        end

        %Drying in grid
        calculateDrying(x,y,C);

        %Return if frosted over
        if didFrost
            out = 1;
            break;
        end
    end
end

function out = calculateDrying(x,y,C,c1,c2,R)
    global frostGrid
    
    R = floor(R);
    for i=-R:R
        for j=-R:R
            
        end
    end
    
    
    out = 1;
end

function frostProb = frostProbability(x,y)
	global frostGrid

    %Check for neightbor frost
    % TODO
    % Check that neighbors are in array
    hasNeighbor = frostGrid(x-1,y +1) == 1 || frostGrid(x,y +1) == 1 ||... 
    frostGrid(x+1,y +1) == 1 || frostGrid(x-1,y) == 1 ||frostGrid(x+1,y) == 1 ||... 
    frostGrid(x-1,y -1) == 1 || frostGrid(x,y-1) == 1 || frostGrid(x+1,y -1) == 1;
    
    if hasNeighbor
        %%have to change
        frostProb = 0.5;
    else
        frostProb = 0;
    end
end

File: Python/Snowman.py
import numpy as np
import random
import math
import matplotlib.pyplot as plt

GRID_SIZE = 151
frostGrid = np.ones([GRID_SIZE,GRID_SIZE], dtype=int)*2
frostGrowth = set()

originX, originY = (GRID_SIZE//2,GRID_SIZE//2)

#Simulation Variables
R = math.sqrt(2)
C = 0.01

frostGrid[originY][originX] = 1
frostGrowth.add((originX,originY))

#walker Variables
walkerRadius = 15
walkerDelta = 1
walkerThreashold = 0.8

running = True
random.seed()

def getFrostProb(x,y):
    hasNeighbor = (frostGrid[y-1][x-1] == 1 or frostGrid[y-1][x] == 1 or
        frostGrid[y-1][x+1] == 1 or frostGrid[y][x-1] == 1 or frostGrid[y][x+1] == 1 or
        frostGrid[y+1][x-1] == 1 or frostGrid[y+1][x] == 1 or frostGrid[y+1][x + 1] == 1)

    if hasNeighbor:
        return 1
    else:
        return 0


def calculateDrying():
    global frostGrid
    c1 = 1
    c2 = 10
    for y in range(1,GRID_SIZE-1):
        for x in range(1,GRID_SIZE-1):
            delta = 0
            if(frostGrid[y][x] == 2):
                if((x+1,y) in frostGrowth) or ((x-1,y) in frostGrowth) or ((x,y+1) in frostGrowth) or ((x,y-1) in frostGrowth):
                    delta = 1
                if ((x+1,y-1) in frostGrowth) or (x-1,y-1) in frostGrowth or (x-1,y+1) in frostGrowth or (x,y+1) in frostGrowth:
                    delta = R
            prob = 0
            if delta != 0:
                prob =  C/(c1*delta+c2)
            if(prob > random.random()):
                frostGrid[y][x] = 0
    draw()
    

def distanceToOriginSq(x,y):
    return (x-originX)**2 + (y-originY)**2

def walk(x,y):
    global walkerRadius
    global running
    global frostGrowth
    ox = x
    oy = y
    maxRestart = 0
    restartCount = 0
    # print("HI")

    while True:
        dir = math.floor(random.uniform(0,4))
        prevX = x
        prevY = y
        # x += random.randint(-1,2)
        # y += random.randint(-1,2)
        if dir ==1:
            x += 1
        elif dir == 2:
            x -= 1
        elif dir == 3:
            y += 1
        else:
            y -= 1

        mag = distanceToOriginSq(x,y)

        # drawCur(x,y)
        
        if mag > walkerRadius**2:
            if restartCount < maxRestart:
                restartCount +=1
                x = ox
                y = oy
            else:
                break
        
        if frostGrid[y][x] != 2:
            if restartCount < maxRestart:
                restartCount +=1
                x = prevX
                y = prevY
            else:
                break

        frostProb = getFrostProb(x,y)
        # print(x,y)

        if random.random() < frostProb:
            # print("stuck")
            frostGrid[y][x] = 1
            frostGrowth.add((x,y))
            calculateDrying()

            if mag >= (walkerThreashold*walkerRadius)**2:
                walkerRadius += walkerDelta
                if walkerRadius >= GRID_SIZE//2:
                    print("PAUSE")
                    running = False
                    walkerRadius -= walkerDelta
                else:
                   print("Radius: ", mag)
            break
        

def drawCur(x,y):
    tmp = frostGrid[y][x]
    frostGrid[y][x] = 3
    plt.imshow(frostGrid, interpolation='none')
    plt.draw()
    plt.pause(0.00001)
    plt.clf()
    frostGrid[y][x] = tmp

def draw(clear = True):
    im = plt.imshow(frostGrid, interpolation='none')
    plt.hot()
    plt.draw()
    plt.colorbar(im)
    plt.pause(0.00001)
    if clear:
        plt.clf()

#Plot
# plt.show(block=False)
itr = 0
while running:
    # print("moving")
    walkerTheta = random.random()*2*math.pi
    # print(walkerTheta)
    walkerX = math.floor(math.cos(walkerTheta)*walkerRadius) + originX
    walkerY = math.floor(math.sin(walkerTheta)*walkerRadius) + originY
    # print(walkerX,walkerY)
    walk(walkerX,walkerY)
    # if itr % 10 == 0:
    #      draw()

    # itr +=1
# draw(clear=False)
im = plt.imshow(frostGrid, interpolation='none')
plt.colorbar(im)
plt.show()
    
